presto-main/src/main/java/com/facebook/presto/sql/planner/LambdaCaptureDesugaringRewriter.java

```diff
@@ -0,0 +1,139 @@
+import com.facebook.presto.sql.tree.BindExpression;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.LambdaArgumentDeclaration;
+import com.facebook.presto.sql.tree.LambdaExpression;
+import com.facebook.presto.sql.tree.SymbolReference;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+import static com.facebook.presto.util.ImmutableCollectors.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public class LambdaCaptureDesugaringRewriter
+{
+    private final Map<Symbol, Type> symbolTypes;
+    private final SymbolAllocator symbolAllocator;
+
+    public LambdaCaptureDesugaringRewriter(Map<Symbol, Type> symbolTypes, SymbolAllocator symbolAllocator)
+    {
+        this.symbolTypes = requireNonNull(symbolTypes, "symbolTypes is null");
+        this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+    }
+
+    public Expression rewrite(Expression expression)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression, new Context());
+    }
+
+    private static Expression replaceSymbols(Expression expression, ImmutableMap<Symbol, Symbol> symbolMapping)
+    {
+        return ExpressionTreeRewriter.rewriteWith(
+                new ExpressionRewriter<Void>() {
+                    @Override
+                    public Expression rewriteSymbolReference(SymbolReference node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+                    {
+                        Symbol mapTo = symbolMapping.get(new Symbol(node.getName()));
+                        if (mapTo == null) {
+                            return node;
+                        }
+                        return mapTo.toSymbolReference();
+                    }
+                },
+                expression);
+    }
+
+    public class Visitor
+            extends ExpressionRewriter<Context>
+    {
+        @Override
+        public Expression rewriteLambdaExpression(LambdaExpression node, Context context, ExpressionTreeRewriter<Context> treeRewriter)
+        {
+            HashSet<Symbol> referencedSymbols = new HashSet<>();
+            Expression rewrittenBody = treeRewriter.rewrite(node.getBody(), context.withReferencedSymbols(referencedSymbols));
+
+            ImmutableList<Symbol> lambdaArguments = node.getArguments().stream()
+                    .map(LambdaArgumentDeclaration::getName)
+                    .map(Symbol::new)
+                    .collect(toImmutableList());
+            referencedSymbols.removeAll(lambdaArguments);
+
+            // x -> f(x, captureSymbol)    will be rewritten into
+            // "$internal$bind"(captureSymbol, (extraSymbol, x) -> f(x, extraSymbol))
```

Comment:
> Bind is fundamentally first class. Just like you wouldn't want to give a type to function application, or treat function application as a normal function, the same cannot be done on bind.

Take a look at Java partial method invocation: https://docs.oracle.com/javase/8/docs/api/java/lang/invoke/MethodHandles.html#insertArguments-java.lang.invoke.MethodHandle-int-java.lang.Object...-
`MethodHandle` definitely has type: https://docs.oracle.com/javase/8/docs/api/java/lang/invoke/MethodHandle.html#type--

Take a look at Ocaml example:
```
# let func x y = x + y;;
val func : int -> int -> int = <fun>
# let bind = func 2;;
val bind : int -> int = <fun>
```
Partial invocation has functional type. Therefore I was expecting something similar from `$internal$bind`.

What does `ExpressionAnalyzer` returns for `$internal$bind`?

> And in the case of Presto, desugaring captures seems to be the most straightforward way to implement captures.

So the main reason is to have simpler, cleaner bytecode generation? +1